app/src/main/java/com/sdp13epfl2021/projmag/database/OfflineCachedCandidatureDatabase.kt

```diff
@@ -0,0 +1,101 @@
+import java.io.File
+import java.io.Serializable
+
+
+private const val CANDIDATURE_FILENAME: String = "candidatures.data"
+
+/**
+ * This is an implementation of CandidatureDatabase that keep candidatures both in persistent storage and memory.
+ * If we are offline and firebase cache is empty/disabled, it will use the data stored locally.
+ * If we are offline and firebase cache is enabled, it will use both the data stored locally and in the cache (priority to `db`).
+ * If we are not offline, it will use the data return by `db` (up to date).
+ */
+class OfflineCachedCandidatureDatabase(
+    private val db: CandidatureDatabase,
+    private val projectsDir: File
+) : CandidatureDatabase {
+
+    private val candidatures: MutableMap<ProjectId, List<Candidature>> = HashMap()
+
+    init {
+        try {
+            projectsDir.listFiles()?.let { it
+                .filter(File::isDirectory)
+                .mapNotNull(File::listFiles)
+                .mapNotNull { arr -> arr.find { f -> f.parentFile != null && f.isFile && f.name == CANDIDATURE_FILENAME } }
+                .forEach { f ->
+                    candidatures[f.parentFile!!.name] = LocalFileUtils.loadFromFile(f, SerializableCandidatureListWrapper::class)?.list ?: emptyList()
+                }
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+    }
+
+    private data class SerializableCandidatureListWrapper(val list: List<Candidature>) :  Serializable
+
+    private fun saveCandidature(projectId: ProjectId) {
+        try {
+            val projectDir: File = File(projectsDir, projectId)
+            projectDir.mkdirs()
+            LocalFileUtils.saveToFile(
+                File(projectDir, CANDIDATURE_FILENAME),
+                SerializableCandidatureListWrapper(candidatures[projectId]!!)
+            )
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+    }
+
+    private fun merge(projectID: ProjectId, remoteList: List<Candidature>): List<Candidature> {
+        val localList = candidatures[projectID]?.let {
+            it.filter {
+                    c -> remoteList.all { r -> r.userID != c.userID }
+            }
+        } ?: emptyList()
+        val totalList = localList + remoteList
+        candidatures[projectID] = totalList
+        saveCandidature(projectID)
+        return totalList
+    }
+
+    override fun getListOfCandidatures(
+        projectID: ProjectId,
+        onSuccess: (List<Candidature>) -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        db.getListOfCandidatures(projectID, { remoteList ->
+            //If we are offline, the remoteList could be empty or incomplete (cache).
+            //We only keep them if they don't collide, with userID (priority to remote => up to date)
+            val totalList: List<Candidature> = merge(projectID, remoteList)
+            onSuccess(totalList)
+        }, onFailure)
+    }
+
```

Comment:
When using the app offline I run into this issue which I assume means the code does  not cache the candidature database offline 
![bug](https://user-images.githubusercontent.com/56798451/117335964-01026f00-ae9c-11eb-89f0-4f30bbf2b37e.png)